File: cpp/784. Letter Case Permutation.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vector<string> ans;
    void dfs(int i, string path, string S) {
        if(i==S.size()) {
            ans.push_back(path);
            return;
        }
        if(isalpha(S[i])) {
            path.push_back(tolower(S[i]));
            dfs(i+1, path, S);
            path.pop_back();
            path.push_back(toupper(S[i]));
            dfs(i+1, path, S);
            path.pop_back();
        } else {
            path.push_back(S[i]);
            dfs(i+1, path, S);
            path.pop_back();
        }
    }
    vector<string> letterCasePermutation(string S) {
        dfs(0, "", S);
        return ans;
    }
};

void printAns(vector<string> ans) {
    for(auto str : ans) {
        cout << str << " ";
    }
    cout << endl;
}

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; string S; vector<string> ans;

    solution = Solution();
    S = "a1b2";
    ans = solution.letterCasePermutation(S);
    printAns(ans);   

    solution = Solution();
    S = "3z4";
    ans = solution.letterCasePermutation(S);
    printAns(ans);   

    solution = Solution();
    S = "12345";
    ans = solution.letterCasePermutation(S);
    printAns(ans);   

    solution = Solution();
    S = "0";
    ans = solution.letterCasePermutation(S);
    printAns(ans);   
}

File: python/784-Letter Case Permutation.py
from typing import List

class Solution:
    def letterCasePermutation(self, S: str) -> List[str]:
        "Backtracking, Time: O(2^n), Space: O(n)"
        result = []
        def dfs(i, path):
            if i == len(S):
                result.append(path)
                return
            if S[i].isalpha():
                dfs(i+1, path+S[i].upper())
                dfs(i+1, path+S[i].lower())
            else:
                dfs(i+1, path+S[i])
        dfs(0, "")
        return result

S = "a1b2"
result = Solution().letterCasePermutation(S)
print(result) # ["a1b2","a1B2","A1b2","A1B2"]

S = "3z4"
result = Solution().letterCasePermutation(S)
print(result) # ["3z4","3Z4"]

S = "12345"
result = Solution().letterCasePermutation(S)
print(result) # ["12345"]

S = "0"
result = Solution().letterCasePermutation(S)
print(result) # ["0"]